deckhouse-controller/pkg/controller/module-controllers/source/helpers.go

```diff
@@ -0,0 +1,276 @@
+	// image digest has 64 symbols, while label can have maximum 63 symbols, so make md5 sum here
+	checksum := fmt.Sprintf("%x", md5.Sum([]byte(meta.Checksum)))
+
+	moduleRelease := &v1alpha1.ModuleRelease{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "ModuleRelease",
+			APIVersion: "deckhouse.io/v1alpha1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: fmt.Sprintf("%s-%s", moduleName, meta.ModuleVersion),
+			Labels: map[string]string{
+				"module":                  moduleName,
+				"source":                  sourceName,
+				"release-checksum":        checksum,
+				release.UpdatePolicyLabel: policy,
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: v1alpha1.ModuleSourceGVK.GroupVersion().String(),
+					Kind:       v1alpha1.ModuleSourceGVK.Kind,
+					Name:       sourceName,
+					UID:        sourceUID,
+					Controller: ptr.To(true),
+				},
+			},
+		},
+		Spec: v1alpha1.ModuleReleaseSpec{
+			ModuleName: moduleName,
+			Version:    semver.MustParse(meta.ModuleVersion),
+			Weight:     meta.ModuleWeight,
+			Changelog:  v1alpha1.Changelog(meta.Changelog),
+		},
+	}
+	if meta.ModuleDefinition != nil {
+		moduleRelease.Spec.Requirements = meta.ModuleDefinition.Requirements
+	}
+
+	if err := r.client.Create(ctx, moduleRelease); err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return err
+		}
+		prevModuleRelease := new(v1alpha1.ModuleRelease)
+		if err = r.client.Get(ctx, client.ObjectKey{Name: moduleRelease.Name}, prevModuleRelease); err != nil {
+			return err
+		}
+
+		// seems weird to update already deployed/suspended release
+		if prevModuleRelease.Status.Phase != v1alpha1.PhasePending {
+			return nil
+		}
+
+		prevModuleRelease.Spec = moduleRelease.Spec
+		return r.client.Update(ctx, prevModuleRelease)
+	}
+	return nil
+}
+
+func (r *reconciler) ensureModule(ctx context.Context, sourceName, moduleName string) (*v1alpha1.Module, error) {
+	module := new(v1alpha1.Module)
+	if err := r.client.Get(ctx, client.ObjectKey{Name: moduleName}, module); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return nil, err
+		}
+		r.log.Debugf("the '%s' module not installed", moduleName)
+		module = &v1alpha1.Module{
+			TypeMeta: metav1.TypeMeta{
+				Kind:       v1alpha1.ModuleGVK.Kind,
+				APIVersion: v1alpha1.ModuleGVK.GroupVersion().String(),
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: moduleName,
+			},
+			Properties: v1alpha1.ModuleProperties{
+				AvailableSources: []string{sourceName},
+			},
+		}
+		r.log.Debugf("the '%s' module not found, create it", moduleName)
+		if err = r.client.Create(ctx, module); err != nil {
+			return nil, err
+		}
+	}
+
+	err := utils.Update[*v1alpha1.Module](ctx, r.client, module, true, func(module *v1alpha1.Module) bool {
+		// init just created downloaded modules
+		if len(module.Status.Conditions) == 0 {
```

Comment:
```suggestion
		if len(module.Status.Conditions) > 0 {
            return false
        }
```

Reply:
as I understand if there are some conditions set, we leave them intact

Reply:
as I understand if there are some conditions set, we leave them intact

Reply:
This part sets conditions to just created modules. Just created modules have no conditions.